### PR TITLE
Add XContent chunking to SearchResponse

### PR DESCRIPTION
**Pull request by romseygeek**
This commit adds xcontent chunking to SearchResponse and MultiSearchResponse
by making SearchHits implement ChunkedToXContent.

Relates to #89838

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/action/search/MultiSearchResponse.java [hunk 1]
```diff
@@ -150,24 +173,14 @@ public void writeTo(StreamOutput out) throws IOException {
...
-            builder.endObject();
-        }
-        builder.endArray();
-        builder.endObject();
-        return builder;
+    public Iterator<? extends ToXContent> toXContentChunked(ToXContent.Params params) {
+        return Iterators.concat(
+            ChunkedToXContentHelper.startObject(),
+            Iterators.single((b, p) -> b.field("took", tookInMillis).startArray(Fields.RESPONSES)),
+            Iterators.flatMap(Arrays.stream(items).iterator(), item -> item.toXContentChunked(params)),
```
Comment (romseygeek) [inline server/src/main/java/org/elasticsearch/action/search/MultiSearchResponse.java:180]: There might be a nicer way to include an array of objects which themselves implement ChunkedToXContent but I couldn't find one

File: server/src/main/java/org/elasticsearch/action/search/SearchResponse.java [hunk 2]
```diff
@@ -264,14 +268,23 @@ public Clusters getClusters() {
...
+            this.innerToXContentChunked(params),
+            ChunkedToXContentHelper.endObject()
+        );
+    }
+
+    public Iterator<? extends ToXContent> innerToXContentChunked(ToXContent.Params params) {
+        return Iterators.concat(
+            Iterators.single((ToXContent) SearchResponse.this::headerToXContent),
+            Iterators.single(clusters),
+            Iterators.flatMap(Iterators.single(internalResponse), r -> r.toXContentChunked(params))
```
Comment (romseygeek) [inline server/src/main/java/org/elasticsearch/action/search/SearchResponse.java:283]: Similarly I feel there ought to be a nicer way of doing this but I couldn't find one...

**Comment by romseygeek** [comment]
This is an interesting test failure.  There's a composite aggs test that checks that we get a specific exception if keys are being formatted in a lossy fashion, and this exception is thrown during xcontent parsing.  But with chunking we don't seem to catch exceptions thrown by parsing code, and so we get a 'failure encoding chunk' message instead.  Is this something we've already encountered elsewhere or do I need to add some extra error handling here?

**Comment by DaveCTurner** [comment]
This is a drawback of the chunked encoding: we have to be sure that the serialization will succeed before we start. Once we start sending chunks to the client we can't our minds about the response code and return an error.

**Comment by romseygeek** [comment]
It looks as though we have quite a few places (mainly in aggs) that have this late checking.  I can work through the tests and try and fix them, but it blocks making any changes here for the moment, unfortunately.

File: server/src/main/java/org/elasticsearch/action/search/MultiSearchResponse.java [hunk 3]
```diff
@@ -82,6 +85,26 @@ public void writeTo(StreamOutput out) throws IOException {
...
 
+        @Override
+        public Iterator<? extends ToXContent> toXContentChunked(ToXContent.Params params) {
+            if (isFailure()) {
+                return Iterators.concat(ChunkedToXContentHelper.startObject(), Iterators.single((b, p) -> {
+                    ElasticsearchException.generateFailureXContent(b, p, Item.this.getFailure(), true);
+                    return b;
+                }),
+                    Iterators.single((b, p) -> b.field(Fields.STATUS, ExceptionsHelper.status(Item.this.getFailure()).getStatus())),
+                    ChunkedToXContentHelper.endObject()
```
Comment (cbuescher) [inline server/src/main/java/org/elasticsearch/action/search/MultiSearchResponse.java:96]: Should this endObject be outside the if-else block? Or removes? The original logic has it in the items loop.
If this is wrong, maybe we need a test to cover this.

File: server/src/main/java/org/elasticsearch/action/search/SearchResponse.java [hunk 4]
```diff
@@ -129,7 +133,7 @@ public SearchResponse(
             : "SearchResponse can't have both scrollId [" + scrollId + "] and searchContextId [" + pointInTimeId + "]";
     }
 
-    @Override
+    // @Override
```
Comment (cbuescher) [inline server/src/main/java/org/elasticsearch/action/search/SearchResponse.java:136]: remove

File: server/src/main/java/org/elasticsearch/action/search/SearchResponseSections.java [hunk 5]
```diff
@@ -98,18 +100,33 @@ public final Map<String, SearchProfileShardResult> profile() {
...
-        }
-        if (profileResults != null) {
-            profileResults.toXContent(builder, params);
-        }
-        return builder;
+    public Iterator<? extends ToXContent> toXContentChunked(ToXContent.Params params) {
+        return Iterators.concat(
+            Iterators.flatMap(Iterators.single(hits), r -> r.toXContentChunked(params)),
+            Iterators.single((ToXContent) (b, p) -> {
+                if (aggregations != null) {
```
Comment (cbuescher) [inline server/src/main/java/org/elasticsearch/action/search/SearchResponseSections.java:107]: I see why moving these checks inside the Iterator helps enumerating everything in one concat call, but this way we create iterators the we already know will be a noop. Don't know if its worth pulling these out. maybe readability suffers then, leave it up to you to decide.

File: server/src/main/java/org/elasticsearch/search/aggregations/metrics/InternalTopHits.java [hunk 6]
```diff
@@ -191,7 +192,7 @@ public Object getProperty(List<String> path) {
 
     @Override
     public XContentBuilder doXContentBody(XContentBuilder builder, Params params) throws IOException {
-        searchHits.toXContent(builder, params);
+        ChunkedToXContent.wrapAsToXContent(searchHits).toXContent(builder, params);
```
Comment (cbuescher) [inline server/src/main/java/org/elasticsearch/search/aggregations/metrics/InternalTopHits.java:195]: Question out of curiosity: I thought the final goal would be to convert everything in search to a "chunked" xcontent variant. This looks a bit like for aggregations we aren't? Is this reading correct, if so is this future work or does that mean aggregations remain larger "chunks"

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/search/action/AsyncSearchResponse.java [hunk 7]
```diff
@@ -194,7 +195,7 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
 
         if (searchResponse != null) {
             builder.field("response");
-            searchResponse.toXContent(builder, params);
+            ChunkedToXContent.wrapAsToXContent(searchResponse).toXContent(builder, params);
```
Comment (cbuescher) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/search/action/AsyncSearchResponse.java:198]: Just want to verify if AsyncSearchResponse will be changed over to ChunkedToXContentObject as well at some point and if we already have plans for that?

**Review by cbuescher — CHANGES_REQUESTED** [review]

**Comment by cbuescher** [comment]
@romseygeek thanks, change looks great in general to me, I left a couple of questions

File: server/src/main/java/org/elasticsearch/action/search/SearchResponse.java (hunk 2 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/action/search/SearchResponse.java:283]: I think you're looking for this:

```suggestion
            internalResponse.toXContentChunked(params)
```

**Comment by cbuescher** [comment]
@romseygeek fyi I'm out for a couple of days, the questions I left aren't strong asks for changes so shouldn't block you from continuing this and for other to approve and get this PR merged.

**Comment by romseygeek** [comment]
@elasticmachine run elasticsearch-ci/part-1

**Comment by romseygeek** [comment]
@elasticmachine run elasticsearch-ci/part-1

File: modules/lang-mustache/src/main/java/org/elasticsearch/script/mustache/SearchTemplateResponse.java [hunk 8]
```diff
@@ -101,16 +102,15 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
         return builder;
     }
 
-    public XContentBuilder innerToXContent(XContentBuilder builder, Params params) throws IOException {
+    public void innerToXContent(XContentBuilder builder, Params params) throws IOException {
```
[the remark this answers is not in the document]
Comment (romseygeek) [reply modules/lang-mustache/src/main/java/org/elasticsearch/script/mustache/SearchTemplateResponse.java:105]: The return value is never used.  Actually I think it makes sense to make this package-private as it's not called from anywhere else either.

File: server/src/main/java/org/elasticsearch/action/search/MultiSearchResponse.java (hunk 3 above)
Comment (romseygeek) [reply server/src/main/java/org/elasticsearch/action/search/MultiSearchResponse.java:95]: The startObject is inside as well, it's just that the code formatting obscures it.  If I change `ElasticsearchException.generateFailureXContent` to return its XContentBuilder then everything fits into one line and is much easier to read, so I'll do that.

File: server/src/main/java/org/elasticsearch/action/search/SearchResponseSections.java (hunk 5 above)
Comment (romseygeek) [reply server/src/main/java/org/elasticsearch/action/search/SearchResponseSections.java:107]: Yeah I went back and forth on this a bit, but I think this is the most readable way of doing it, and creating no-op iterators is pretty low cost.

File: server/src/main/java/org/elasticsearch/search/aggregations/metrics/InternalTopHits.java (hunk 6 above)
Comment (romseygeek) [reply server/src/main/java/org/elasticsearch/search/aggregations/metrics/InternalTopHits.java:195]: Yes, the plan is to do this bit-by-bit.  So we start off by chunking search into search hits and aggregations, and we can then look at breaking these chunks into smaller ones in the future.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/search/action/AsyncSearchResponse.java (hunk 7 above)
Comment (romseygeek) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/search/action/AsyncSearchResponse.java:198]: I've created #95661 to keep track of these.

**Comment by romseygeek** [comment]
This is ready for another round of reviews

**Comment by romseygeek** [comment]
@elasticmachine update branch

File: server/src/main/java/org/elasticsearch/search/fetch/FetchPhase.java [hunk 9]
```diff
@@ -72,8 +74,12 @@ public void execute(SearchContext context) {
 
         Profiler profiler = context.getProfilers() == null ? Profiler.NOOP : Profilers.startProfilingFetchPhase();
         SearchHits hits = null;
-        try {
-            hits = buildSearchHits(context, profiler);
+        try (XContentBuilder xContentValidator = new XContentBuilder(XContentType.JSON.xContent(), OutputStream.nullOutputStream())) {
```
Comment (cbuescher) [inline server/src/main/java/org/elasticsearch/search/fetch/FetchPhase.java:77]: Question for better understanding: this validator builder is used to completely serialize the search hits once on the node before sending them across the wire and then thrown away? Is there an understanding of the extra work this involves? I see why validation is necessary for chunking but I want to make sure I understand the tradeoffs here.

**Review by cbuescher — COMMENTED** [review]
Thanks for the update, I left one question regarding how much extra work we expect for the xContent validation happening for the search hits in the FetchPhase.

File: server/src/main/java/org/elasticsearch/search/fetch/FetchPhase.java (hunk 9 above)
Comment (romseygeek) [reply server/src/main/java/org/elasticsearch/search/fetch/FetchPhase.java:77]: Actually I think #95673 means that I can remove this bit, good catch!

**Review by cbuescher — APPROVED** [review]
Thanks, LGTM

**Pull request merged**